### PR TITLE
dbeaver/dbeaver#34398 Fix resultset interval ordering for Postgres

### LINKED ISSUE #34398: ORDER BY fails on "interval" PG datatype
### Description

When right-clicking a column of PG "interval" datatype and using GUI's ORDER BY, DBeaver orders alphabetically. It results in a different order than PG's native ORDER BY.
image

### DBeaver Version

Community Edition 23.3.2

### Operating System

Windows 11

### Database and driver

PostgreSQL 15.3.0

### Steps to reproduce

1. 
  ~~~sql
CREATE TABLE test (c1 INTERVAL)

INSERT INTO test (c1)
VALUES (INTERVAL '9 days') ,(INTERVAL '1 year')

SELECT *
FROM test
ORDER BY c1
  ~~~
2. Right-click on the c1 column
3. Order / ORDER BY c1 ASC
4. Witness that DBeaver considers that "1 year" is less than "9 days"

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by HocKu7**

### CODE REVIEW

File: plugins/org.jkiss.dbeaver.ext.postgresql/src/org/jkiss/dbeaver/ext/postgresql/model/data/PostgreIntervalValueHandler.java [hunk 1]
```diff
@@ -97,4 +117,50 @@ public String getValueDisplayString(@NotNull DBSTypedObject column, Object value
...
+            if (days != null) {
+                totalSeconds += Long.parseLong(days) * MILLISECONDS_IN_DAY;
+            }
+            if (time != null) {
+                String[] timeParts = time.split(":");
+                totalSeconds += (long) Integer.parseInt(timeParts[0]) * MILLISECONDS_IN_HOUR;
+                totalSeconds += (long) Integer.parseInt(timeParts[1]) * MILLISECONDS_IN_MINUTE;
+                long intValue = (long) Double.parseDouble(timeParts[2]);
+                totalSeconds += intValue * MILLISECONDS_IN_SECOND;
+                totalSeconds += (long) (Double.parseDouble(timeParts[2]) - intValue);
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.ext.postgresql/src/org/jkiss/dbeaver/ext/postgresql/model/data/PostgreIntervalValueHandler.java:155]: `java.time.Duration` please? :^)

File: plugins/org.jkiss.dbeaver.ext.postgresql/src/org/jkiss/dbeaver/ext/postgresql/model/data/PostgreIntervalValueHandler.java [hunk 2]
```diff
@@ -44,6 +47,23 @@ public class PostgreIntervalValueHandler extends JDBCStringValueHandler {
...
+    private static final long MILLISECONDS_IN_MONTH = 30 * MILLISECONDS_IN_DAY;
+    private static final long MILLISECONDS_IN_YEAR = 365 * MILLISECONDS_IN_DAY;
+
+    // parsing values like: 3 mons 15 days, 00:00:00.12, -9 days
+    private final String intervalRegex = "(?i)(?<sign>-)?\\s*" +
+        "(?:(?<years>\\d+)\\s+years?)?\\s*" +
+        "(?:(?<months>\\d+)\\s+mon(?:s|ths)?)?\\s*" +
+        "(?:(?<days>\\d+)\\s+days?)?\\s*" +
+        "(?:(?<time>\\d{2}:\\d{2}:\\d{2}(?:\\.\\d+)?))?\\b";
+    private final Pattern pattern = Pattern.compile(intervalRegex);
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.ext.postgresql/src/org/jkiss/dbeaver/ext/postgresql/model/data/PostgreIntervalValueHandler.java]: Please embed regex inside the pattern and make it static. If you want to leave the regex string separately for aesthetic purposes, please move it to the static constructor.

File: plugins/org.jkiss.dbeaver.ext.postgresql/src/org/jkiss/dbeaver/ext/postgresql/model/data/PostgreIntervalValueHandler.java [hunk 3]
```diff
@@ -97,4 +117,50 @@ public String getValueDisplayString(@NotNull DBSTypedObject column, Object value
...
+    @Override
+    public Comparator<Object> getComparator() {
+        return (o1, o2) -> {
+            Long leftInterval = getSecondsFromInterval((String) o1);
+            Long rightInterval = getSecondsFromInterval((String) o2);
+            return Long.compare(leftInterval, rightInterval);
+        };
+    }
+
+    private Long getSecondsFromInterval(String interval) {
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.ext.postgresql/src/org/jkiss/dbeaver/ext/postgresql/model/data/PostgreIntervalValueHandler.java]: Why `Long`?

File: plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/model/data/DBDValueHandler.java [hunk 4]
```diff
@@ -117,4 +119,13 @@ Object createNewValueObject(@NotNull DBCSession session, @NotNull DBSTypedObject
...
     DBCLogicalOperator[] getSupportedOperators(@NotNull DBSTypedObject attribute);
 
+    /**
+     * Get comparator for compare values
+     *
+     * @return comparator
+     */
+    default Comparator<Object> getComparator() {
+        return null;
+    }
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/model/data/DBDValueHandler.java:129]: Let's not extend the basic interface with such a crude functionality (I don't like it's constrained to `Object`). Instead, let the implementations implement `Comparable` as well.

File: plugins/org.jkiss.dbeaver.ui.editors.data/src/org/jkiss/dbeaver/ui/controls/resultset/ResultSetModel.java [hunk 5]
```diff
@@ -1138,7 +1138,10 @@ public void resetOrdering() {
                     }
                     Object cell1 = getCellValue(new ResultSetCellLocation(binding, row1));
                     Object cell2 = getCellValue(new ResultSetCellLocation(binding, row2));
-                    if (cell1 instanceof String && cell2 instanceof String) {
+                    Comparator<Object> comparator = columnElement.getValueHandler().getComparator();
+                    if (comparator != null) {
+                        result = comparator.compare(cell1, cell2);
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.ui.editors.data/src/org/jkiss/dbeaver/ui/controls/resultset/ResultSetModel.java:1143]: ... then you can do

```java
if (columnElement.getValueHandler() instanceof Comparable<?>) { ... }
```

**Review by ShadelessFox — CHANGES_REQUESTED** [review]

**Comment by uslss** [comment]
verified

**Pull request merged**